### PR TITLE
[TT-6927] Improving DX with make lint, add faillint, PR template update

### PR DESCRIPTION
**Pull request by titpetric**
Essentially this PR improves a number of things in regards to the developer experience. There are a number of scenarios which hurt our delivery due to time wasted on CI and a very long feedback loop, and often actions which should be first-order development tasks before opening a PR.

## Description

- Added `make lint`, updated CI to use it. People don't run shit on local, I hope this will change this.
- Added `goimports` and `gofmt` as part of `make lint` (make people run it on local)
- Added .faillint to blacklist `fmt.Print*` (use our /log pkg) and `net/http.Get` (no timeouts).
- Added goenv `.go-version`, we build against 1.16 - facilitate people actually using 1.16 with goenv.
- Updated PR template (removed unused/obsolete checklists) - most of them are handled in CI, we don't need to double-blind the inputs. I've left only the essential ones that apply.
- If a new field is added into the /config/config.go, usually a separate linter surfaces an issue in the CI. The linter is now added into `make lint`.
- Decreased verbosity on failing tests, scrolling through 100 pages of content has no value. This PR removes spam both in github comments and emails.

## Related Issue

As of yet - none. This falls under maintenance and CI, DX, and should only be merged to the develop branch, or particularly the latest release branch. Backporting it isn't required as it fixes no immediate issues, only improves ongoing development.

## Motivation and Context

Happyness.

## How This Has Been Tested

Lots of `make lint` invocations. GH actions have been left to CI.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Refactoring or add test (improvements in base code or adds test coverage to functionality)

## Checklist

- [x] I ensured that the documentation is up to date
- [x] I explained why this PR updates go.mod in detail with reasoning why it's required
- [x] I would like a code coverage CI quality gate exception and have explained why

This is a maintenance PR. There are no implementations added for which code coverage should be addressed. There are no go.mod changes.

### CODE REVIEW

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4314/merge
Commit:  
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/3132563316)

File: apidef/oas/middleware.go [hunk 1]
```diff
@@ -71,15 +75,37 @@ func (g *Global) ExtractTo(api *apidef.APIDefinition) {
...
-	AllowCredentials   bool     `bson:"allowCredentials,omitempty" json:"allowCredentials,omitempty"`
-	ExposedHeaders     []string `bson:"exposedHeaders,omitempty" json:"exposedHeaders,omitempty"`
-	AllowedHeaders     []string `bson:"allowedHeaders,omitempty" json:"allowedHeaders,omitempty"`
-	OptionsPassthrough bool     `bson:"optionsPassthrough,omitempty" json:"optionsPassthrough,omitempty"`
-	Debug              bool     `bson:"debug,omitempty" json:"debug,omitempty"`
-	AllowedOrigins     []string `bson:"allowedOrigins,omitempty" json:"allowedOrigins,omitempty"`
-	AllowedMethods     []string `bson:"allowedMethods,omitempty" json:"allowedMethods,omitempty"`
+	// If set to `true`, this option turns CORS processing on or off.
+	// Old API definition: `CORS.enable`.
+	Enabled bool `bson:"enabled" json:"enabled"` // required
```
Comment (jeffy-mathew) [inline apidef/oas/middleware.go:80]: It'll be better to have the godoc pattern used here right? Comment starting with field name

File: apidef/oas/root.go [hunk 2]
```diff
@@ -44,19 +44,22 @@ func (x *XTykAPIGateway) ExtractTo(api *apidef.APIDefinition) {
...
 	// Name is the name of the API.
-	// Old API Definition: `name`
-	Name       string      `bson:"name" json:"name"` // required
-	Expiration string      `bson:"expiration,omitempty" json:"expiration,omitempty"`
-	State      State       `bson:"state" json:"state"` // required
+	// Old API definition: `name`
+	Name string `bson:"name" json:"name"` // required
+	// Expiration date.
+	Expiration string `bson:"expiration,omitempty" json:"expiration,omitempty"`
+	// API state configuration.
```
Comment (jeffy-mathew) [inline apidef/oas/root.go:60]: godoc patter

**Comment by jeffy-mathew** [comment]
We had a discussion around what we should call the Tyk API definition - whether it be `old API definition` `Tyk Native` API definition, and we decided to go with `Tyk native API definition`. Can we consider this as an opportunity to change the name `old API definition` to `Tyk native API definition` in the comments and thereby generated docs? @titpetric 
cc: @lghiur @furkansenharputlu

**Comment by ermirizio** [comment]
@titpetric CI changes looks good for me, the addition of the linters directly into the makefile makes much more sense so this can be exercised locally as well before push.

The following is a suggestion for taking a look, into how to split large PRs for "ease" the review process: https://blog.logrocket.com/using-stacked-pull-requests-in-github/ .
I think in this case we have a bunch of changes ( 1- CI changes, 2- Linter changes, 3- Code changes itself) and those could fit different squads for review as well, let me know your thoughts about this, I think you could find it useful, credits to @alephnull .
In this case approving the changes for the core codebase itself is fine for me , but I don't feel this is all within my expertise, I mean I don't want to fall into some blind approval behaviour, so stacked PRs could help to sort this situations

**Comment by titpetric** [comment]
@ermirizio correct, already in the process to split the PR due to TT-5744. On vacay, so it's pretty low time dedication on my side, given the places I go to 😂

The stdlogger removal feels like unjustified, it's a breaking change just because it removes old trash code. I have to start deprecating stuff instead, so that change goes into dev/null, leaving only the make lint changes here. Will remove -v from that one test (lower verbosity = better).

Wouldn't split the PR template 🤷‍♂️ but not strongly opinionated about it

File: .go-version [hunk 3]
```diff
@@ -0,0 +1 @@
+1.16.0
```
[the remark this answers is not in the document]
Comment (titpetric) [reply .go-version:1]: > Added goenv .go-version, we build against 1.16 - facilitate people actually using 1.16 with goenv.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 4]
```diff
@@ -1,46 +1,44 @@
...
 ## Related Issue
-<!-- This project only accepts pull requests related to open issues -->
-<!-- If suggesting a new feature or change, please discuss it in an issue first -->
-<!-- If fixing a bug, there should be an issue describing it with steps to reproduce -->
-<!-- Please link to the issue here -->
+
+<!-- This project only accepts pull requests related to open issues. -->
+<!-- If suggesting a new feature or change, please discuss it in an issue first. -->
+<!-- If fixing a bug, there should be an issue describing it with steps to reproduce. -->
+<!-- OSS: Please link to the issue here. Tyk: please create/link the JIRA ticket. -->
```
[the remark this answers is not in the document]
Comment (titpetric) [reply .github/PULL_REQUEST_TEMPLATE.md:12]: Changed as agreed upon, moved to separate PR :)

File: apidef/oas/middleware.go (hunk 1 above)
Comment (titpetric) [reply apidef/oas/middleware.go]: For sure. Added godoc comments for complete oas package in #4327, moved everything there.

**Comment by titpetric** [comment]
@ermirizio i feel the git-state output could be improved. How do we feel about git diff per-file? The report is coming from a goimports violation on master.

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4314/merge
Commit: 6f7c88ca42fb552b4c8c88b901f2d539b0f13cd7 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/3184042887)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4314/merge
Commit: 5c932d2c199eabd4ca38847556ee91691d1077fd 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/3184103577)

**Comment by ermirizio** [comment]
> @ermirizio i feel the git-state output could be improved. How do we feel about git diff per-file? The report is coming from a goimports violation on master.

basically you are printing the git diff-index output after the lint process which exposes issues not being addressed while locally developing.. 

Adding more verbosity (git diff per file) could be a nice to have, however I consider more beneficial that we could open a PR automatically if make lint detects an issue, what do you think about?

**Review by ermirizio — APPROVED** [review]
Besides my comments this looks good for me, good additions for the linting subject @titpetric

File: Makefile [hunk 5]
```diff
@@ -21,6 +21,24 @@ BENCH_RUN=NONE
 test:
 	$(GOTEST) -run=$(TEST_REGEX) -count=$(TEST_COUNT) ./...
 
+# lint runs all local linters that must pass before pushing
+.PHONY: lint lint-install lint-fast
+lint: lint-install
+	goimports -w .
```
Comment (mitjaziv) [inline Makefile:27]: `goimports` and `gofmt` are already part of `golangci-lint` does it make sense to run it separately?

https://golangci-lint.run/usage/linters/

You would reduce amount with 3rd party requirements when using only `golangci-lint` and you would reduce setup time, locally and in CI.

File: Makefile [hunk 6]
```diff
@@ -21,6 +21,24 @@ BENCH_RUN=NONE
...
+	gofmt -w .
+	faillint -ignore-tests -paths "$(shell grep -v '^#' .faillint | xargs echo | sed 's/ /,/g')" ./...
+	go mod tidy
```
Comment (mitjaziv) [inline Makefile:30]: I don't think `go mod tidy` is part of linting actions but if you add it, I think it should be run before linters or I think linters will download stuff.

**Review by jeffy-mathew — APPROVED** [review]

File: Makefile (hunk 5 above)
Comment (titpetric) [reply Makefile:27]: It makes sense, yes, but with some more concerns on the config. Will be a separate PR adding goci lint.

**Review by tbuchaillot — APPROVED** [review]

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4314/merge
Commit: 9d6c7696824ce0088d9626fe3d62c227a55ae257 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/3345942290)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4314/merge
Commit: be00317f7cb72ff4b7ac326005ca7ac1b6c98872 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/3345988903)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4314/merge
Commit: e802406644fee9d832e9bd39760d03158cdc9eaa 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/3346079509)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4314/merge
Commit: 251de0ded258f31e82e550b4147a21c58f1b5402 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/3346114729)

**Comment by titpetric** [comment]
```
$ git cherry-pick f964ae4740237ef0eddaf7b8aa90f4c67611f073
Auto-merging .github/workflows/ci-tests.yml
Auto-merging Makefile
Auto-merging cli/importer/importer.go
Auto-merging gateway/mw_graphql_test.go
[release-4 72c8a365] [TT-6927] Improving DX with make lint, add faillint, PR template update (#4314)
 Date: Fri Oct 28 16:41:36 2022 +0200
 13 files changed, 70 insertions(+), 55 deletions(-)
 create mode 100644 .faillint
 create mode 100644 .go-version
```

There didn't seem to be any conflict.

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/heads/master
Commit: f964ae4740237ef0eddaf7b8aa90f4c67611f073 [TT-6927] Improving DX with make lint, add faillint, PR template update (#4314)

* Add make lint command, update PR template, update ci-tests to use make lint

* Add -count 1 lines to prevent caching

* Fix goimports for analytics_test.go, broken on master

* Move go mod tidy before running linters

* Print full git diff for git-state

* Fix goimports issues on master
Triggered by: push (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/3346283075)

**Pull request merged**